include/thread_pool/thread_pool.hpp

```diff
@@ -169,10 +169,15 @@ inline bool GenericThreadPool<Task, Queue>::tryPost(Handler&& handler)
         auto result = m_idle_workers.tryEmptyAny();
         if (result.first)
         {
-            auto success = m_workers[result.second]->tryPost(std::forward<Handler>(handler));
-            m_workers[result.second]->wake();
-
-            return success;
+            if (m_workers[result.second]->tryPost(std::forward<Handler>(handler)))
+            {
+                m_workers[result.second]->wake();
+                return true;
+            }
+
+            // If post is unsuccessful, we need to re-add the worker to the idle worker bag.
+            m_idle_workers.fill(result.second);
```

Comment:
is there a race here?  should we not let the thread add itself back once it is awake?

Reply:
This should be safe, the previous version had the thread add itself back in, this just short circuits the condition variable overhead (see https://github.com/SynaptiveMedical/thread-pool-cpp/pull/3#discussion_r174533662).